### ISSUE TITLE
Having a miserable time with Chimp in CircleCI.

### ISSUE THREAD

**Issue by martinhbramwell** [xolvio/chimp#443]
It seems evident that Meteor is running far too slowly.

I'm hoping these logs can help find the reason

#####  Expected behaviour

On development machine Chimp Cucumber returns : 

    25 scenarios (25 passed)
    103 steps (103 passed)
    2m02.112s 

#####  Actual behaviour

On CircleCI Chimp Cucumber returns : 

    25 scenarios (8 failed, 17 passed)
    103 steps (8 failed, 6 skipped, 89 passed)
    3m29.485s

The first sign of trouble : 

    [chimp][phantom.stderr] 2016-08-05T12:54:22 [CRITICAL] QNetworkReplyImpl: backend error: caching was enabled after some bytes had been written

    [chimp][phantom.stdout] [ERROR - 2016-08-05T12:54:23.380Z] Session [bb140c80-5b0b-11e6-a443-655aaf578ba0] - page.onError - msg: TypeError: undefined is not an object (evaluating 'Package.modules.meteorInstall')

    [chimp][phantom.stdout]   phantomjs://platform/console++.js:263 in error
    [ERROR - 2016-08-05T12:54:23.380Z] Session [bb140c80-5b0b-11e6-a443-655aaf578ba0] - page.onError - stack:
      (anonymous function) (http://localhost:3000/packages/es5-shim.js?hash=e4d68c9ebbcdbd230d16f4bd933ee8dbf2978a22:17)

      phantomjs://platform/console++.js:263 in error
    
That is then followed by . . .  

    [chimp] Detected an unhandledRejection:
    [chimp][hooks] Could not provide error hint
    [chimp][hooks] Promise: { exception: 
       { message: 'Promise was rejected with the following reason: timeout',
         type: 'WaitUntilTimeoutError' } }
    [chimp][hooks] Forcibly exiting Cucumber

    [chimp][cucumber] Received message from cucumber child. Result: {"message":"Promise was rejected with the following reason: timeout","type":"WaitUntilTimeoutError"}

. . . and also : 

    [chimp] Detected an unhandledRejection:
    [chimp][hooks] TypeError: Cannot read property 'getFileName' of undefined
        at isFrameInCucumber (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/cucumber/lib/cucumber/runtime/stack_trace_filter.js:9:23)
        at Array.<anonymous> (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/cucumber/lib/cucumber/runtime/stack_trace_filter.js:15:9)
        at TraceModifier._modify (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/stack-chain/stack-chain.js:44:32)
        at Function.prepareStackTrace (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/stack-chain/stack-chain.js:119:25)
        at trackRejection (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/xolvio-sync-webdriverio/node_modules/webdriverio/node_modules/q/q.js:1066:32)
        at reject (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/xolvio-sync-webdriverio/node_modules/webdriverio/node_modules/q/q.js:1131:5)
        at deferred.reject (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/xolvio-sync-webdriverio/node_modules/webdriverio/node_modules/q/q.js:633:16)
        at run (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/babel-runtime/node_modules/core-js/library/modules/es6.promise.js:87:22)
        at /opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/babel-runtime/node_modules/core-js/library/modules/es6.promise.js:100:28
        at flush (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/babel-runtime/node_modules/core-js/library/modules/_microtask.js:18:9)
        at nextTickCallbackWith0Args (node.js:420:9)
        at process._tickCallback (node.js:349:13)
    [chimp][hooks] Promise: { _c: [], _a: [], _s: 2, _d: true, _v: {}, _h: 0, _n: false }
    [chimp][hooks] Forcibly exiting Cucumber

##### Steps to reproduce the problem

The repo is :   [warehouseman/meteor-mantra-kickstarter :: branch = ci_trial](https://github.com/warehouseman/meteor-mantra-kickstarter/tree/ci_trial)

#####  Used version

    * Chimp: 0.39.3
    * Node.js: v4.4.7
    * Operation system: Ubuntu 14.04
    * java version "1.8.0_40"
    Java(TM) SE Runtime Environment (build 1.8.0_40-b26)
    Java HotSpot(TM) 64-Bit Server VM (build 25.40-b25, mixed mode)
    _JAVA_OPTIONS: -Xms512m -Xmx1024m
    
#####  Console / Log Output

[Chimp log with --debug ](https://gist.github.com/martinhbramwell/832f0fa6396c07fe797f898d61e97fa0)

------- 
*Join our Slack [xolv.io/community](http://xolv.io/community) #chimp channel, where you can find help and help others.*

**Comment by lgandecki** [comment]
It works for us with over thousand of scenarios. It is much slower in comparison to our dev machines, but the virtual machines there are much slower... We've had a lot of problems with tests failing on circle while working on local, but we've fixed them one by one. They all turned out to be timing issues, caused by, again, the fact that circle vms were slower. You could try circle enterprise if you want to use circle environment but have better times.

**Comment by martinhbramwell** [comment]
So, it seems you are suggesting stuffing in waits just ahead of every error I encounter.

The problem is the error of type : 

    [chimp][hooks] Promise: { exception: 
       { message: 'Promise was rejected with the following reason: timeout',
         type: 'WaitUntilTimeoutError' }  

How do I increase the timeout on a Promise?

**Comment by martinhbramwell** [comment]
It seems I should re-word the question to :  "How do I determine which timeout failed, and how do I increase it?"

**Comment by lgandecki** [comment]
To the contrary - we've had quite a bit of waits all around that were the cause of the timeouts - the wait time was enough for desktops but not for circle. We've switched to more dynamic timeouts - waiting untill something happens, instead of waiting a specific amount of time etc. 

One thing, I'd suggest using chrome for testing on circle - phantomjs is not very reliable This change might just automagically solve your problems.

Another - getFileName is this somewhere in your code?

You can also set it up to run in parallel, and then run separate scenarios in separate VMs, then you will know which ones are failing. The log you show here is weird though, I do agree with that. Are there not surrounded by specific scenarios/steps? that would help to see which steps are failing and why

**Comment by martinhbramwell** [comment]
> We've switched to more dynamic timeouts - waiting untill something happens, instead of waiting a specific amount of time etc. 

Yes I learned that years ago with straight Selenium; as much as possible I wait for elements to be enabled.  The deliberate 3 minute wait, before starting Chimp, was in the CircleCi bash script.

> One thing, I'd suggest using chrome for testing on circle - phantomjs is not very reliable This change might just automagically solve your problems.

Yeah, I'll try that right now.

>  Another - getFileName is this somewhere in your code?

No, I have no idea what it is, or how I could find out. I grep'd my whole repo and did not find it. Googling it turns up nothing either!   Is it NOT part of Chimp? I'm going to assume it's in PhantomJs. :-)

> The log you show here is weird though, I do agree with that. Are there not surrounded by specific scenarios/steps? that would help to see which steps are failing and why

Well that's run with --debug on Chimp, so you see way too much.  The scenarios are there, but not easy to see for all the noise.

**Comment by samhatoum** [comment]
@martinhbramwell did you get this resolved? I would like to close this as it's a discussion and not really an issue. Better suited for forums and stackoverflow.

Thanks!

**Comment by martinhbramwell** [comment]
I believe I have tracked down the issue : 

    [chimp] Detected an unhandledRejection:
    [chimp][hooks] TypeError: Cannot read property 'getFileName' of undefined
    at isFrameInCucumber     (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/cucumber/lib/cucumber/runtime/stack_trace_filter.js:9:23)
    at Array.<anonymous> (/opt/circleci/nodejs/v4.4.7/lib/node_modules/chimp/node_modules/cucumber/lib/cucumber/runtime/stack_trace_filter.js:15:9)

It's not an issue with Chimp *per se*, but it is a solvable problem when using Chimp in a slow environment.

I'll report on the solution later today or tomorrow . . .

**Comment by martinhbramwell** [comment]
I did these three things to get past the problem : 

##### 1) Get past the, "TypeError: Cannot read property 'getFileName' of undefined" problem.

It seems there is no default, (or the default in CircleCI is unusable), for the destination path of WebdriverIO execution logs.  So I run Chimp with the following command.  Note that the config file must be specified first! : 

    chimp .e2e_tests/chimp-config.js --ddp=http://localhost:3000 --path=.e2e_tests

The [target directory](https://github.com/warehouseman/meteor-mantra-kickstarter/blob/v0.2.1/.e2e_tests/chimp-config.js#L7) is specified in `.e2e_tests/chimp-config.js`, which looks like this ; 

    module.exports = {
      // - - - - WEBDRIVER-IO - - - -
      webdriverio: {
        logOutput: '/tmp/logs/meteor/'
      },
    };

##### 2) Get past the, "Promise was rejected with the following reason: timeout" problem.

The lack of a place to write log files was obscuring the underlying timing problems.  Moreover, there were two timing problems, one concealing the other, I believe.  The obvious one is solved with explicit waits such as, [waitForEnabled](http://webdriver.io/api/utility/waitForEnabled.html).

The problem with `waitForEnabled`, and similar "wait" methods, is that the retest interval is only `500msecs`, with no option to increase it.  So, if CircleCI's container is **not** ready for the next try within that ½ second, it tries to write an error to `logOutput`.

For those cases, I used [waitUntil](http://webdriver.io/api/utility/waitUntil.html), which **does** allow explicit interval settings.

##### 3) Get more consistent behavior.

I'm not certain it made any difference, but I stopped requiring PhantomJS, and let Chimp run the default.

**Comment by jirikrepl** [comment]
Hello I think you can setup waitforTimeout of waitForXXX functions by this config http://webdriver.io/guide/getstarted/configuration.html#waitforTimeout

```
module.exports = {
    ...
    webdriverio: {
        waitforTimeout: 10000,
        ...
    },
};
```
